include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+// Currently supports variants of type: int
+template <typename T, std::enable_if_t<std::is_same_v<T, int>, bool> = true>
+class KnobSlider final
+    : public Knob
+{
+public:
+    KnobSlider(const std::string& flagName, T defaultValue, T minValue, T maxValue)
+        : Knob(flagName)
+    {
+        PPX_ASSERT_MSG(minValue < maxValue, "invalid range to initialize slider");
+        PPX_ASSERT_MSG(minValue <= defaultValue && defaultValue <= maxValue, "defaultValue is out of range");
+
+        mMinValue = minValue;
+        mMaxValue = maxValue;
+        SetDefault(defaultValue);
+    }
+
+    void Draw() override
+    {
+        ImGui::SliderInt(mDisplayName.c_str(), &mValue, mMinValue, mMaxValue, NULL, ImGuiSliderFlags_AlwaysClamp);
+    }
+
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // GetFlagHelpText format:
+    // --flag_name <mMinValue~mMaxValue> : flag description
+    std::string GetFlagHelpText() const override
+    {
+        std::string flagHelpText = "--" + mFlagName + " <" + std::to_string(mMinValue) + "~" + std::to_string(mMaxValue) + ">";
+        if (!mFlagDesc.empty()) {
+            flagHelpText += " : " + mFlagDesc;
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <int>
+    void UpdateFromFlag(const CliOptions& opts)
+    {
+        SetDefault(opts.GetExtraOptionValueOrDefault(mFlagName, mValue));
+    }
+
+    T GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // Updates mValue
+    void SetValue(T newVal)
+    {
+        if (!IsValidValue(newVal)) {
+            PPX_LOG_ERROR(mFlagName << " cannot be set to " << newVal << " because it's out of range " << mMinValue << "~" << mMaxValue);
+            return;
+        }
+
+        mValue = newVal;
+    }
+
+private:
+    // IsValidValue returns true if val is within allowed range
+    bool IsValidValue(T val)
+    {
+        if (val < mMinValue || val > mMaxValue)
+            return false;
+        return true;
+    }
+
+    // SetDefault will set mDefaultValue and update mValue
+    void SetDefault(T newVal)
+    {
+        PPX_ASSERT_MSG(IsValidValue(newVal), "invalid default value");
+        mDefaultValue = newVal;
+        ResetToDefault();
+    }
+
+private:
+    T mValue;
+    T mDefaultValue;
+
+    // mValue will be clamped to the mMinValue to mMaxValue range, inclusive
+    T mMinValue;
+    T mMaxValue;
+};
+
+// KnobDropdown will be displayed as a dropdown in the UI
+// The knob stores the index of a selected choice from a list of allowed options
+// Currently supports option variants of type: string
+template <typename T, std::enable_if_t<std::is_same_v<T, std::string>, bool> = true>
+class KnobDropdown final
+    : public Knob
+{
+public:
+    KnobDropdown(const std::string& flagName, size_t defaultIndex, std::vector<T>::iterator choicesBegin, std::vector<T>::iterator choicesEnd)
```

Comment:
Using `std::vector<T>::iterator` is very constraining. If you don't feel confortable using iterators, prefer switching back to a const ref of a container, but as-is, I think we get the worse of both cases:

containers over iterators:
- pro: easy to read
- pro: ability to use inline constructs like `foo({value1, value2})`
- con: have to use this specific container, if API chooses vector, vector it is. (Except if you go with @hysw's solution, which is to template the container which is a bit more flexible).

iterators:
- con: a bit harder to read if not used to
- con: cannot use the inline construct AFAIK
- pro: not bound to a container at all. Input can be anything as long as you implement the iterator constraints.

Here, by using `std::vector<T>::iterator`, you force the use of a non-constant iterator coming from `std::vector`
 - cannot use cbegin/cend
 - vector has to be non-const
 
By using `std::input_iterator auto choicesBegin`, you allow any iterator as long as they at least have the constraints listed by `std::input_iterator`, which are:
 - you can increment the iterator to go to the next item
 - you can read it.
Meaning it could be an iterator to a tree, a constant array, anything, as long as you can read and it has the logic for the increment operation.

Reply:
Or just have both:
```
template<typename T>
KnobDropdown(const std::string& flagName, size_t defaultIndex, std::input_iterator<T> choicesBegin, std::input_iterator<T> choicesEnd) : ... mChoices(choicesBegin, choicesEnd) {}
template<typename T>
KnobDropdown(const std::string& flagName, size_t defaultIndex, const T& container) : KnobDropdown(flagName, defaultIndex, std::begin(container), std::end(container))
```

Reply:
Yes, sure. Just not `std::vector::iterator` :)